### PR TITLE
Enhance tests for files API root best-effort pin.

### PR DESCRIPTION
**Pull request by kevina**
License: MIT
Signed-off-by: Kevin Atkinson <k@kevina.org>

### CODE REVIEW

**Comment by whyrusleeping** [comment]
Sweet, LGTM!

@jbenet wanna take a look here?

File: test/sharness/t0252-files-gc.sh [hunk 1]
```diff
@@ -42,4 +43,34 @@ test_expect_success "gc okay after adding incomplete node" '
...
+'
+
+test_expect_success "run gc and make sure directory contents are removed" '
+  ipfs repo gc &&
+  test_must_fail ipfs cat $HELLO_HASH
+'
+
+test_expect_success "add incomplete directory and make sure gc is okay" '
+  ipfs files cp /ipfs/$MYDIR_HASH /mydir &&
+  ipfs repo gc
```
Comment (jbenet) [inline test/sharness/t0252-files-gc.sh]: I would do `ipfs cat $HELLO_HASH` here too, just to be sure the thing is really gone and stays gone.

File: test/sharness/t0252-files-gc.sh [hunk 2]
```diff
@@ -42,4 +43,34 @@ test_expect_success "gc okay after adding incomplete node" '
...
 
+MYDIR_HASH=QmRLrxQTdmLZNfDJxFQ9uFrgtj4s6cZU9GzTxCq2YLev24
+HELLO_HASH=QmeV1kwh3333bsnT6YRfdCRrSgUPngKmAhhTa4RrqYPbKT
+
+test_expect_success "add directory with direct pin" '
+  mkdir mydir/ &&
+  echo "hello world!" > mydir/hello.txt &&
+  ipfs add --pin=false -r mydir &&
+  ipfs pin add --recursive=false $MYDIR_HASH &&
+  ipfs cat $HELLO_HASH
```
Comment (jbenet) [inline test/sharness/t0252-files-gc.sh]: if something changes, the hashes may no longer match watch you got. if it's important that it matches, check it explicitly. if it is not important that it matches, grab the hashes directly from the output of `ipfs add` with something like:

```sh
HELLO_HASH=$(ipfs add --pin=false -q -r mydir/hello.txt | tail -n1)
MYDIR_HASH=$(ipfs add --pin=false -q -r mydir | tail -n1)
```

File: test/sharness/t0252-files-gc.sh [hunk 3]
```diff
@@ -42,4 +43,34 @@ test_expect_success "gc okay after adding incomplete node" '
   ipfs refs $ADIR_HASH
 '
 
+MYDIR_HASH=QmRLrxQTdmLZNfDJxFQ9uFrgtj4s6cZU9GzTxCq2YLev24
+HELLO_HASH=QmeV1kwh3333bsnT6YRfdCRrSgUPngKmAhhTa4RrqYPbKT
```
Comment (jbenet) [inline test/sharness/t0252-files-gc.sh]: i would name these more usefully:
- `DIR_TO_STAY` or `DIR_PINNED`
- `FILE_TO_GO` or `FILE_UNPINNED`

File: test/sharness/t0252-files-gc.sh [hunk 4]
```diff
@@ -11,13 +11,15 @@ test_description="test how the unix files api interacts with the gc"
...
-  echo "hello world"  | ipfs files write --create /hello.txt &&
+  echo "hello world" > hello.txt
+  cat hello.txt | ipfs files write --create /hello.txt &&
   ipfs repo gc &&
   ipfs cat QmVib14uvPnCP73XaCDpwugRuwfTsVbGyWbatHAmLSdZUS
 '
 
 test_expect_success "/hello.txt still accessible after gc" '
-  ipfs files read /hello.txt
+  ipfs files read /hello.txt > hello-actual
```
Comment (jbenet) [inline test/sharness/t0252-files-gc.sh]: need `&&` at the end

File: test/sharness/t0252-files-gc.sh [hunk 5]
```diff
@@ -11,13 +11,15 @@ test_description="test how the unix files api interacts with the gc"
 test_init_ipfs
 
 test_expect_success "object not removed after gc" '
-  echo "hello world"  | ipfs files write --create /hello.txt &&
+  echo "hello world" > hello.txt
```
Comment (jbenet) [inline test/sharness/t0252-files-gc.sh]: need `&&` at the end

**Comment by jbenet** [comment]
Some comments above

File: test/sharness/t0252-files-gc.sh (hunk 5 above)
Comment (kevina) [inline test/sharness/t0252-files-gc.sh]: will fix

File: test/sharness/t0252-files-gc.sh (hunk 4 above)
Comment (kevina) [inline test/sharness/t0252-files-gc.sh]: will fix

File: test/sharness/t0252-files-gc.sh (hunk 3 above)
Comment (kevina) [inline test/sharness/t0252-files-gc.sh]: Will use DIR_PINNED and FILE_UNPINNED.

File: test/sharness/t0252-files-gc.sh (hunk 1 above)
Comment (kevina) [inline test/sharness/t0252-files-gc.sh]: Will add.

File: test/sharness/t0252-files-gc.sh (hunk 2 above)
Comment (kevina) [inline test/sharness/t0252-files-gc.sh]: Sounds like a good idea will do.

**Comment by kevina** [comment]
@jbenet thanks, I just amended my commit with your corrections.

Sorry about the missing "&&", I actually meant to review this one more time before removing the "in progress" label.

**Comment by whyrusleeping** [comment]
This LGTM. I would change a few stylistic things (like redirect the 'ipfs cat' calls either to dev/null, or check their output explicitly). But thats not a huge deal, and isnt really relevant to what we're testing here.

**Pull request merged**